In [33]:
# Fix for OpenBLAS issue with scikit-learn
import os
os.environ['OMP_NUM_THREADS'] = '1'
os.environ['OPENBLAS_NUM_THREADS'] = '1'

In [34]:
pip install pandas sqlalchemy psycopg2-binary


Note: you may need to restart the kernel to use updated packages.


In [35]:
import pandas as pd

In [36]:
from sqlalchemy import create_engine

# Update connection string if needed
engine = create_engine("postgresql+psycopg2://postgres:9899@localhost:5432/postgres")


In [37]:
q = """
WITH base AS (
  SELECT
    fs.year_month,
    fs.customer_id,
    c.customer_name,
    c.sales_agent_id,
    c.loyalty_member,
    c.digital_tool_user,
    fs.category,
    fs.discount_rate,
    fs.net_sales_eur
  FROM analytics.fact_sales fs
  JOIN analytics.dim_customer c
    ON fs.customer_id = c.customer_id
  WHERE fs.country = 'DE'
    AND c.customer_type = 'Independent Optician'
    AND fs.year_month BETWEEN '2024-01' AND '2024-08'
)
SELECT
  year_month,
  customer_id,
  customer_name,
  sales_agent_id,
  loyalty_member,
  digital_tool_user,
  category,
  discount_rate,
  net_sales_eur
FROM base;
"""

df = pd.read_sql(q, engine)



In [38]:
print("Min month:", df["year_month"].min())
print("Max month:", df["year_month"].max())

print("\nAll months:")
print(sorted(df["year_month"].astype(str).unique()))

print("\nCustomers:", df["customer_id"].nunique())
print("Rows:", len(df))


Min month: 2024-01
Max month: 2024-08

All months:
['2024-01', '2024-02', '2024-03', '2024-04', '2024-05', '2024-06', '2024-07', '2024-08']

Customers: 87
Rows: 1740


In [39]:
#STEP 2 — Understand Buying Frequency

#How regular is buying behavior in this customer base?


active_months = (
    df.groupby("customer_id")["year_month"]
      .nunique()
      .reset_index(name="active_month_count")
)

print(active_months.describe())
print("\nDistribution:")
print(active_months["active_month_count"].value_counts().sort_index())
active_months.to_excel("active_customers_monthly.xlsx", index=False)


       active_month_count
count           87.000000
mean             5.068966
std              1.198770
min              2.000000
25%              4.000000
50%              5.000000
75%              6.000000
max              8.000000

Distribution:
2     1
3     5
4    25
5    24
6    21
7    10
8     1
Name: active_month_count, dtype: int64


Customers are:

👉 Mostly regular buyers
👉 Not opportunistic one-off buyers
👉 Not seasonal-only buyers
👉 Not disappearing quickly

Most are active 50–75% of the year.

In [40]:
active_trend = (
    panel.groupby("year_month")["active"]
    .sum()
    .reset_index(name="active_customers")
)

active_trend.to_csv("active_customers_trend.csv", index=False)


In [41]:
# STEP 3 — Build Monthly Customer Panel

monthly = (
    df.groupby(["customer_id", "year_month"])
      .agg(
          net_sales=("net_sales_eur", "sum"),
          avg_discount=("discount_rate", "mean"),
          category_count=("category", "nunique"),
          loyalty=("loyalty_member", "max"),
          digital=("digital_tool_user", "max"),
          sales_agent_id=("sales_agent_id", "max")
      )
      .reset_index()
)

print("Monthly rows:", len(monthly))
print("Unique customers:", monthly["customer_id"].nunique())
print("Unique months:", monthly["year_month"].nunique())


Monthly rows: 441
Unique customers: 87
Unique months: 8


On average we have: 441 / 8 = 55 active customers per month

In [42]:
# STEP 4 — Create Full Customer × Month Grid

import pandas as pd

# Ensure month is datetime
monthly["year_month"] = pd.to_datetime(monthly["year_month"])

# Create full month range
all_months = pd.date_range(
    monthly["year_month"].min(),
    monthly["year_month"].max(),
    freq="MS"
)

customers = monthly["customer_id"].unique()

# Create grid
grid = pd.MultiIndex.from_product(
    [customers, all_months],
    names=["customer_id", "year_month"]
).to_frame(index=False)

panel = grid.merge(monthly,
                   on=["customer_id", "year_month"],
                   how="left")

# Fill missing months with zeros
panel["net_sales"] = panel["net_sales"].fillna(0)
panel["avg_discount"] = panel["avg_discount"].fillna(0)
panel["category_count"] = panel["category_count"].fillna(0)
panel["loyalty"] = panel["loyalty"].fillna(0)
panel["digital"] = panel["digital"].fillna(0)

panel["active"] = (panel["net_sales"] > 0).astype(int)

print("Panel rows:", len(panel))
print("Active months:", panel["active"].sum())
print("Inactive months:", len(panel) - panel["active"].sum())


Panel rows: 696
Active months: 441
Inactive months: 255


The average customer is active 63% of the time.
441 / 696 = 0.63

In [43]:
# STEP 5 — Define 2-Month Dormancy (Confirmed)
# We need t+1 and t+2 months to confirm a 2-month inactivity event.
# Since the dataset ends in Aug 2024, the latest month we can confirm is Jun 2024.

panel = panel.sort_values(["customer_id", "year_month"]).copy()

cutoff_month = pd.Timestamp("2024-06-01")
panel_churn = panel[panel["year_month"] <= cutoff_month].copy()
panel_churn = panel_churn.sort_values(["customer_id", "year_month"]).copy()

panel_churn["next1"] = panel_churn.groupby("customer_id")["active"].shift(-1)
panel_churn["next2"] = panel_churn.groupby("customer_id")["active"].shift(-2)

# Only months where we can observe the full 2-month look-ahead are eligible
panel_churn["eligible_2m"] = panel_churn["next2"].notna()

panel_churn["dormancy_2m_confirmed"] = (
    panel_churn["eligible_2m"] &
    (panel_churn["active"] == 1) &
    (panel_churn["next1"] == 0) &
    (panel_churn["next2"] == 0)
).astype(int)

# Calculate dormancy rate on eligible, active rows only (correct denominator)
eligible_active = panel_churn[panel_churn["eligible_2m"] & (panel_churn["active"] == 1)]
dormancy_rate = eligible_active["dormancy_2m_confirmed"].mean()

print("2-month dormancy rate (confirmed, through Jun 2024):", dormancy_rate)
print("Total dormancy events:", int(panel_churn["dormancy_2m_confirmed"].sum()))
print("Eligible active rows:", len(eligible_active))



2-month dormancy rate (confirmed, through Jun 2024): 0.13636363636363635
Total dormancy events: 30
Eligible active rows: 220


Among active-month observations, about 10% resulted in a 2-month inactivity sequence.


In [44]:
# STEP 6 — How Many Unique Customers Became Dormant (2M Confirmed)
unique_dormant_customers = panel_churn.loc[
    panel_churn["dormancy_2m_confirmed"] == 1, "customer_id"
].nunique()

print("Unique customers flagged as 2M dormant (confirmed):", unique_dormant_customers)



Unique customers flagged as 2M dormant (confirmed): 30


That’s:

Approximately 49% of customers experienced one two-month inactivity episode during the 8-month window, reflecting moderate purchasing volatility rather than permanent attrition.

In [45]:
# STEP 7 — Revenue at Risk (Confirmed Window)
# Event-month revenue for confirmed 2M dormancy events
revenue_at_risk = panel_churn.loc[
    panel_churn["dormancy_2m_confirmed"] == 1, "net_sales"
].sum()

total_revenue_window = panel_churn["net_sales"].sum()

print("Revenue at risk (event-month, confirmed):", revenue_at_risk)
print("Total revenue (Jan–Jun window):", total_revenue_window)
print("Share of revenue at risk (within window):", revenue_at_risk / total_revenue_window)



Revenue at risk (event-month, confirmed): 20923.499999999996
Total revenue (Jan–Jun window): 253431.22
Share of revenue at risk (within window): 0.08256086207531967


In [46]:
# Revenue from customers who had at least one confirmed dormancy event (within Jan–Jun window)
dormant_customers = panel_churn.loc[
    panel_churn["dormancy_2m_confirmed"] == 1, "customer_id"
].unique()

revenue_from_dormant_customers = panel_churn.loc[
    panel_churn["customer_id"].isin(dormant_customers),
    "net_sales"
].sum()

print("Revenue from dormant-flagged customers (Jan–Jun):", revenue_from_dormant_customers)



Revenue from dormant-flagged customers (Jan–Jun): 65629.97


In [47]:
# Revenue split by dormancy flag (event-month rows, Jan–Jun)
panel_churn.groupby("dormancy_2m_confirmed")["net_sales"].sum()



dormancy_2m_confirmed
0    232507.72
1     20923.50
Name: net_sales, dtype: float64

10% of total revenue is associated with customers who experienced at least one 2-month inactivity episode.
This is not 10% of revenue permanently lost.
It is 10% of revenue tied to customers who had confirmed dormancy at some point.


In [48]:
# STEP 8 — Compare Dormant vs Stable (Eligible Active Rows Only)
analysis = (
    panel_churn[panel_churn["eligible_2m"] & (panel_churn["active"] == 1)]
    .groupby("dormancy_2m_confirmed")
    .agg(
        avg_revenue=("net_sales", "mean"),
        median_revenue=("net_sales", "median"),
        avg_discount=("avg_discount", "mean"),
        avg_category=("category_count", "mean"),
        loyalty_rate=("loyalty", "mean"),
        digital_rate=("digital", "mean"),
        count=("customer_id", "count")
    )
)

print(analysis)



                       avg_revenue  median_revenue  avg_discount  \
dormancy_2m_confirmed                                              
0                       700.586737         606.245      0.198462   
1                       697.450000         531.310      0.211679   

                       avg_category  loyalty_rate  digital_rate  count  
dormancy_2m_confirmed                                                   
0                          1.831579      0.536842      0.484211    190  
1                          1.900000      0.500000      0.400000     30  


* Churned customers are NOT low-value customers. They behave similarly to stable customers before dormancy.
* Discount level does not appear to be a differentiating factor between stable and churned customers.
* Product breadth is not predictive either. Churned customers are not less diversified.

Dormancy events are not driven by revenue size, discount levels, or category diversification. The portfolio exhibits moderate cyclical volatility rather than structural weakness.

In [49]:
# STEP 9 — Agent Risk (Confirmed Window)
agent_risk = (
    panel_churn[panel_churn["dormancy_2m_confirmed"] == 1]
    .groupby("sales_agent_id")
    .agg(
        dormancy_events=("customer_id", "count"),
        revenue_at_risk=("net_sales", "sum")
    )
    .reset_index()
)

agent_base = (
    panel_churn[panel_churn["eligible_2m"] & (panel_churn["active"] == 1)]
    .groupby("sales_agent_id")
    .agg(
        total_eligible_active_events=("customer_id", "count"),
        total_revenue=("net_sales", "sum")
    )
    .reset_index()
)

agent_summary = agent_base.merge(agent_risk, on="sales_agent_id", how="left").fillna(0)

agent_summary["dormancy_rate"] = (
    agent_summary["dormancy_events"] / agent_summary["total_eligible_active_events"]
)

agent_summary["revenue_risk_share"] = (
    agent_summary["revenue_at_risk"] / agent_summary["total_revenue"]
)

print(agent_summary.sort_values("revenue_at_risk", ascending=False))



   sales_agent_id  total_eligible_active_events  total_revenue  \
1            A004                            21       12658.30   
7            A014                            11        8550.98   
5            A009                            11        7895.99   
11           A018                            29       20246.86   
3            A006                            17       16759.07   
2            A005                            20       12209.18   
6            A011                            22       15285.57   
0            A001                            11        9607.73   
4            A007                            16        7302.52   
10           A017                            13        8052.84   
8            A015                            27       17105.97   
9            A016                            22       18359.97   

    dormancy_events  revenue_at_risk  dormancy_rate  revenue_risk_share  
1               7.0          4927.70       0.333333            0.38

In [50]:
# STEP 10 — Revenue Concentration
# Total revenue per customer (Jan–Aug)
customer_revenue = (
    panel.groupby("customer_id")["net_sales"]
         .sum()
         .sort_values(ascending=False)
)

total_revenue = customer_revenue.sum()

# Calculate cumulative revenue share
cumulative_share = customer_revenue.cumsum() / total_revenue

# Add rank and share into dataframe
pareto_df = pd.DataFrame({
    "customer_id": customer_revenue.index,
    "total_revenue": customer_revenue.values,
    "cumulative_share": cumulative_share.values
})

pareto_df["customer_rank"] = range(1, len(pareto_df)+1)

print(pareto_df.head(10))

# Top 20% customers
top_20_count = int(0.2 * len(pareto_df))
top_20_share = pareto_df.iloc[top_20_count-1]["cumulative_share"]

print("\nTop 20% customer revenue share:", top_20_share)


  customer_id  total_revenue  cumulative_share  customer_rank
0      C00089        9285.92          0.027624              1
1      C00002        7970.37          0.051334              2
2      C00219        7718.83          0.074297              3
3      C00106        7706.94          0.097223              4
4      C00173        6942.43          0.117876              5
5      C00126        6898.69          0.138398              6
6      C00178        6875.86          0.158853              7
7      C00062        6635.04          0.178591              8
8      C00082        6100.81          0.196740              9
9      C00080        5765.33          0.213890             10

Top 20% customer revenue share: 0.32756064957640757


In [51]:
# Top customers by confirmed dormancy-event revenue (Jan–Jun)
risk_customers = (
    panel_churn[panel_churn["dormancy_2m_confirmed"] == 1]
    .groupby("customer_id")
    .agg(
        revenue_at_risk=("net_sales", "sum"),
        last_active_month=("year_month", "max"),
        sales_agent_id=("sales_agent_id", "max")
    )
    .reset_index()
    .sort_values("revenue_at_risk", ascending=False)
)

print(risk_customers.head(10))



   customer_id  revenue_at_risk last_active_month sales_agent_id
29      C00220          2106.18        2024-01-01           A009
27      C00206          1725.48        2024-04-01           A004
10      C00081          1582.36        2024-04-01           A018
19      C00138          1511.39        2024-03-01           A014
18      C00135          1246.69        2024-03-01           A014
9       C00051          1137.71        2024-04-01           A009
17      C00134          1104.21        2024-04-01           A006
14      C00107           990.32        2024-01-01           A005
1       C00007           910.01        2024-03-01           A006
23      C00160           844.63        2024-03-01           A004


In [52]:
# Customers active in June
june_customers = set(
    panel[(panel["year_month"] == "2024-06-01") & (panel["active"] == 1)]["customer_id"]
)

# Customers active in July
july_customers = set(
    panel[(panel["year_month"] == "2024-07-01") & (panel["active"] == 1)]["customer_id"]
)

# Dropped from June to July
dropped = june_customers - july_customers

print("June active:", len(june_customers))
print("July active:", len(july_customers))
print("Dropped June→July:", len(dropped))


June active: 60
July active: 45
Dropped June→July: 33


In [53]:
june_revenue = panel[
    (panel["year_month"] == "2024-06-01") &
    (panel["customer_id"].isin(dropped))
]["net_sales"].sum()

print("Revenue lost June→July:", june_revenue)


Revenue lost June→July: 31990.260000000002


In [54]:
monthly_revenue = (
    panel.groupby("year_month")["net_sales"]
    .sum()
    .reset_index()
)

print(monthly_revenue)


  year_month  net_sales
0 2024-01-01   31282.20
1 2024-02-01   31593.58
2 2024-03-01   40141.52
3 2024-04-01   51017.68
4 2024-05-01   40338.15
5 2024-06-01   59058.09
6 2024-07-01   32395.81
7 2024-08-01   50327.42


In [55]:
# Customers active in June but not July
dropped_customers = june_customers - july_customers

# Check how many months those customers were active overall
dropped_activity = active_months[
    active_months["customer_id"].isin(dropped_customers)
]

print(dropped_activity["active_month_count"].describe())
print(dropped_activity["active_month_count"].value_counts().sort_index())


count    33.000000
mean      4.787879
std       0.992395
min       3.000000
25%       4.000000
50%       5.000000
75%       5.000000
max       7.000000
Name: active_month_count, dtype: float64
3     1
4    15
5     9
6     6
7     2
Name: active_month_count, dtype: int64


In [56]:
# Calculate average monthly revenue per customer (Jan–May baseline)
baseline = (
    panel[
        (panel["year_month"] < "2024-06-01") &
        (panel["customer_id"].isin(dropped_customers))
    ]
    .groupby("customer_id")["net_sales"]
    .mean()
    .reset_index(name="baseline_avg")
)

# June revenue for dropped customers
june_rev = (
    panel[
        (panel["year_month"] == "2024-06-01") &
        (panel["customer_id"].isin(dropped_customers))
    ][["customer_id", "net_sales"]]
    .rename(columns={"net_sales": "june_revenue"})
)

comparison = baseline.merge(june_rev, on="customer_id")

comparison["spike_ratio"] = comparison["june_revenue"] / comparison["baseline_avg"]

print(comparison["spike_ratio"].describe())


count    33.000000
mean      2.497852
std       2.645719
min       0.168070
25%       1.136333
50%       1.765890
75%       2.965201
max      14.722623
Name: spike_ratio, dtype: float64


In [57]:
# Spike ratio table for dropped cohort (Power BI export)

import pandas as pd
import numpy as np

# Ensure year_month is datetime
panel["year_month"] = pd.to_datetime(panel["year_month"])

june = pd.Timestamp("2024-06-01")
july = pd.Timestamp("2024-07-01")

# dropped cohort = June active but July inactive
june_customers = set(panel[(panel["year_month"] == june) & (panel["active"] == 1)]["customer_id"])
july_customers = set(panel[(panel["year_month"] == july) & (panel["active"] == 1)]["customer_id"])
dropped_customers = june_customers - july_customers

# Baseline avg Jan–May for dropped cohort
baseline = (
    panel[(panel["year_month"] < june) & (panel["customer_id"].isin(dropped_customers))]
    .groupby("customer_id")["net_sales"]
    .mean()
    .reset_index(name="baseline_avg_jan_may")
)

# June revenue for dropped cohort
june_rev = (
    panel[(panel["year_month"] == june) & (panel["customer_id"].isin(dropped_customers))]
    .groupby("customer_id")["net_sales"]
    .sum()
    .reset_index(name="june_revenue")
)

# Add agent + flags
meta = (
    panel[(panel["year_month"] == june) & (panel["customer_id"].isin(dropped_customers))]
    .groupby("customer_id")
    .agg(
        sales_agent_id=("sales_agent_id","max"),
        loyalty=("loyalty","max"),
        digital=("digital","max")
    )
    .reset_index()
)

spike = baseline.merge(june_rev, on="customer_id", how="inner").merge(meta, on="customer_id", how="left")

# Protect against divide-by-zero
spike["spike_ratio"] = np.where(
    spike["baseline_avg_jan_may"] > 0,
    spike["june_revenue"] / spike["baseline_avg_jan_may"],
    np.nan
)

# Optional: cap extreme ratios for chart readability
spike["spike_ratio_capped"] = spike["spike_ratio"].clip(upper=10)

spike = spike.sort_values("spike_ratio", ascending=False)

spike.to_csv("SpikeRatio_DroppedCohort.csv", index=False)

print("Rows:", len(spike))
print(spike["spike_ratio"].describe())


Rows: 33
count    33.000000
mean      2.497852
std       2.645719
min       0.168070
25%       1.136333
50%       1.765890
75%       2.965201
max      14.722623
Name: spike_ratio, dtype: float64


* Median = 1.77 > Half of dropped customers spent at least 77% more than normal in June.
* 75% > 2.97 = 25% of dropped customers spent nearly 3x baseline in June.
* July inactivity is largely consistent with post-spike normalization.

In [58]:
monthly_revenue = (
    df.groupby("year_month")["net_sales_eur"]
      .sum()
      .reset_index()
      .sort_values("year_month")
)

print(monthly_revenue)


  year_month  net_sales_eur
0    2024-01       31282.20
1    2024-02       31593.58
2    2024-03       40141.52
3    2024-04       51017.68
4    2024-05       40338.15
5    2024-06       59058.09
6    2024-07       32395.81
7    2024-08       50327.42


In [59]:
# Customers active in June but not July
june_customers = set(
    panel[(panel["year_month"] == "2024-06-01") & (panel["active"] == 1)]["customer_id"]
)

july_customers = set(
    panel[(panel["year_month"] == "2024-07-01") & (panel["active"] == 1)]["customer_id"]
)

dropped = june_customers - july_customers

# Who returned in August?
august_customers = set(
    panel[(panel["year_month"] == "2024-08-01") & (panel["active"] == 1)]["customer_id"]
)

recovered = dropped & august_customers

print("Dropped in July:", len(dropped))
print("Recovered in August:", len(recovered))
print("Recovery rate:", len(recovered) / len(dropped))


Dropped in July: 33
Recovered in August: 23
Recovery rate: 0.696969696969697


In [60]:
august_revenue_recovered = panel[
    (panel["year_month"] == "2024-08-01") &
    (panel["customer_id"].isin(dropped))
]["net_sales"].sum()

print("Recovered revenue in August:", august_revenue_recovered)


Recovered revenue in August: 15801.35


In [61]:
import pandas as pd

june = pd.Timestamp("2024-06-01")
july = pd.Timestamp("2024-07-01")
aug = pd.Timestamp("2024-08-01")

# June active customers
june_active = set(
    panel[
        (panel["year_month"] == june) &
        (panel["active"] == 1)
    ]["customer_id"]
)

# July active customers
july_active = set(
    panel[
        (panel["year_month"] == july) &
        (panel["active"] == 1)
    ]["customer_id"]
)

# August active customers
aug_active = set(
    panel[
        (panel["year_month"] == aug) &
        (panel["active"] == 1)
    ]["customer_id"]
)

# Dropped in July
dropped_july = june_active - july_active

# Still inactive in August
true_risk_customers = dropped_july - aug_active

print("True non-recovered customers:", len(true_risk_customers))
print(true_risk_customers)


True non-recovered customers: 10
{'C00157', 'C00080', 'C00086', 'C00091', 'C00155', 'C00217', 'C00108', 'C00106', 'C00073', 'C00167'}


In [62]:
risk_table = (
    panel[
        (panel["customer_id"].isin(true_risk_customers)) &
        (panel["year_month"] == june)
    ][["customer_id", "sales_agent_id", "net_sales"]]
    .rename(columns={"net_sales": "last_june_revenue"})
    .sort_values("last_june_revenue", ascending=False)
)

print(risk_table)


    customer_id sales_agent_id  last_june_revenue
221      C00080           A015            1976.74
317      C00106           A011            1539.70
253      C00086           A017            1254.21
677      C00217           A015            1212.22
477      C00157           A009            1106.23
333      C00108           A017             665.69
269      C00091           A018             537.46
205      C00073           A005             496.98
469      C00155           A015             272.82
501      C00167           A015             129.48


In [63]:
import numpy as np
concentration_trend = []

for month in sorted(panel["year_month"].unique()):
    
    month_data = panel[
        (panel["year_month"] == month) &
        (panel["net_sales"] > 0)   # <-- CRITICAL FIX
    ]

    customer_rev = (
        month_data.groupby("customer_id")["net_sales"]
        .sum()
        .sort_values(ascending=False)
    )

    total_rev = customer_rev.sum()
    
    if total_rev == 0:
        continue

    active_customers = len(customer_rev)

    top_20_n = int(np.ceil(0.2 * active_customers))

    top_20_share = customer_rev.iloc[:top_20_n].sum() / total_rev

    concentration_trend.append({
        "year_month": month,
        "active_customers": active_customers,
        "total_revenue": total_rev,
        "top20_share": top_20_share
    })

concentration_df = pd.DataFrame(concentration_trend)

print(concentration_df)
concentration_df.to_csv("top20_share_trend.csv", index=False)



  year_month  active_customers  total_revenue  top20_share
0 2024-01-01                54       31282.20     0.488071
1 2024-02-01                52       31593.58     0.412094
2 2024-03-01                59       40141.52     0.399616
3 2024-04-01                55       51017.68     0.389139
4 2024-05-01                53       40338.15     0.420569
5 2024-06-01                60       59058.09     0.428048
6 2024-07-01                45       32395.81     0.372866
7 2024-08-01                63       50327.42     0.441200


In [64]:
import numpy as np
import pandas as pd

# Use full panel for general KPIs
panel_local = panel.copy()
panel_local["year_month"] = pd.to_datetime(panel_local["year_month"])

# Bring confirmed 2M dormancy flag into full panel (NaN where not eligible / not confirmed)
flag_cols = panel_churn[["customer_id", "year_month", "eligible_2m", "dormancy_2m_confirmed"]].copy()
flag_cols["year_month"] = pd.to_datetime(flag_cols["year_month"])

panel_local = panel_local.merge(
    flag_cols,
    on=["customer_id", "year_month"],
    how="left"
)

# Monthly KPI summary
kpi_monthly = []

for month in sorted(panel_local["year_month"].unique()):
    month_data = panel_local[panel_local["year_month"] == month]

    active_customers = int(month_data["active"].sum())
    total_revenue = float(month_data["net_sales"].sum())
    rolling_rate = month_data["active_3m"].mean() if "active_3m" in month_data else None

    # Confirmed 2M dormancy rate is only defined where eligible_2m is True
    eligible_active = month_data[(month_data.get("eligible_2m", False) == True) & (month_data["active"] == 1)]
    if len(eligible_active) > 0:
        dormancy_rate = eligible_active["dormancy_2m_confirmed"].mean()
    else:
        dormancy_rate = np.nan

    # Concentration (Top 20% share among active customers)
    active_only = month_data[month_data["net_sales"] > 0]
    cust_rev = active_only.groupby("customer_id")["net_sales"].sum().sort_values(ascending=False)

    if len(cust_rev) > 0:
        top_n = int(np.ceil(0.2 * len(cust_rev)))
        top_share = float(cust_rev.iloc[:top_n].sum() / cust_rev.sum())
    else:
        top_share = 0.0

    kpi_monthly.append({
        "year_month": month,
        "total_revenue": total_revenue,
        "active_customers": active_customers,
        "rolling_3m_active_rate": rolling_rate,
        "top20_share": top_share,
        "dormancy_2m_rate_confirmed": dormancy_rate
    })

kpi_df = pd.DataFrame(kpi_monthly)
kpi_df.to_csv("monthly_kpis.csv", index=False)

print(kpi_df)



  year_month  total_revenue  active_customers rolling_3m_active_rate  \
0 2024-01-01       31282.20                54                   None   
1 2024-02-01       31593.58                52                   None   
2 2024-03-01       40141.52                59                   None   
3 2024-04-01       51017.68                55                   None   
4 2024-05-01       40338.15                53                   None   
5 2024-06-01       59058.09                60                   None   
6 2024-07-01       32395.81                45                   None   
7 2024-08-01       50327.42                63                   None   

   top20_share  dormancy_2m_rate_confirmed  
0     0.488071                    0.166667  
1     0.412094                    0.076923  
2     0.399616                    0.203390  
3     0.389139                    0.090909  
4     0.420569                         NaN  
5     0.428048                         NaN  
6     0.372866                      

In [65]:
import pandas as pd
import numpy as np
from pathlib import Path

out_dir = Path("powerbi_exports")
out_dir.mkdir(exist_ok=True)

# ----------------------------
# 1) CustomerMonth fact (panel)
# ----------------------------
customer_month = panel.copy()
customer_month["year_month"] = pd.to_datetime(customer_month["year_month"])

# keep only columns Power BI needs
customer_month = customer_month[
    ["customer_id","year_month","net_sales","avg_discount","category_count",
     "loyalty","digital","sales_agent_id","active"]
].copy()

customer_month.to_csv(out_dir / "CustomerMonth.csv", index=False)

# -----------------------------------------
# 2) Dormancy events (confirmed through Jun)
# -----------------------------------------
dormancy_events = panel_churn.copy()
dormancy_events["year_month"] = pd.to_datetime(dormancy_events["year_month"])

dormancy_events = dormancy_events[
    dormancy_events["eligible_2m"] == True
].copy()

# event rows only (flag=1) + a version including non-events if you want denominators in BI
dormancy_events_all = dormancy_events[
    ["customer_id","year_month","sales_agent_id","active","eligible_2m",
     "dormancy_2m_confirmed","net_sales","avg_discount","category_count","loyalty","digital"]
].copy()
dormancy_events_all.to_csv(out_dir / "Dormancy2M_Eligible_All.csv", index=False)

dormancy_events_flagged = dormancy_events_all[dormancy_events_all["dormancy_2m_confirmed"] == 1].copy()
dormancy_events_flagged.rename(columns={"net_sales":"event_month_revenue"}, inplace=True)
dormancy_events_flagged.to_csv(out_dir / "Dormancy2M_Events.csv", index=False)

# ----------------------------
# 3) Monthly KPI table
# ----------------------------
# (if you already built kpi_df in the notebook, reuse it; otherwise compute quick)
kpi_df.to_csv(out_dir / "MonthlyKPIs.csv", index=False)

# ----------------------------
# 4) Customer Pareto totals
# ----------------------------
pareto_export = pareto_df.copy()
pareto_export.to_csv(out_dir / "CustomerPareto.csv", index=False)

# ----------------------------
# 5) June→July drop cohort
# ----------------------------
june = pd.Timestamp("2024-06-01")
july = pd.Timestamp("2024-07-01")
aug = pd.Timestamp("2024-08-01")

june_active = customer_month[(customer_month["year_month"] == june) & (customer_month["active"] == 1)][
    ["customer_id","sales_agent_id","net_sales"]
].rename(columns={"net_sales":"june_sales"})

july_active_ids = set(
    customer_month[(customer_month["year_month"] == july) & (customer_month["active"] == 1)]["customer_id"]
)
aug_active_ids = set(
    customer_month[(customer_month["year_month"] == aug) & (customer_month["active"] == 1)]["customer_id"]
)

june_active["dropped_in_july"] = ~june_active["customer_id"].isin(july_active_ids)
june_active["recovered_in_aug"] = june_active["customer_id"].isin(aug_active_ids)

drop_cohort = june_active[june_active["dropped_in_july"] == True].copy()
drop_cohort["true_non_recovered"] = ~drop_cohort["recovered_in_aug"]

drop_cohort.to_csv(out_dir / "DropCohort_JunToJul.csv", index=False)

# ----------------------------
# 6) Agent summaries (ready KPI tables)
# ----------------------------
agent_dormancy_summary = (
    dormancy_events_all[dormancy_events_all["active"] == 1]
    .groupby("sales_agent_id")
    .agg(
        eligible_active_events=("customer_id","count"),
        total_revenue=("net_sales","sum"),
        dormancy_events=("dormancy_2m_confirmed","sum"),
        revenue_at_risk=("net_sales", lambda x: float(x[dormancy_events_all.loc[x.index,"dormancy_2m_confirmed"]==1].sum()))
    )
    .reset_index()
)

agent_dormancy_summary["dormancy_rate"] = agent_dormancy_summary["dormancy_events"] / agent_dormancy_summary["eligible_active_events"]
agent_dormancy_summary["revenue_risk_share"] = agent_dormancy_summary["revenue_at_risk"] / agent_dormancy_summary["total_revenue"]

agent_dormancy_summary.to_csv(out_dir / "AgentDormancySummary.csv", index=False)


print(f"✅ Exports written to: {out_dir.resolve()}")
    

NameError: name 'agent_july_summary' is not defined

In [66]:
import pandas as pd
import numpy as np

panel = panel.copy()
panel["year_month"] = pd.to_datetime(panel["year_month"])
panel = panel.sort_values(["customer_id","year_month"])


In [67]:
june = pd.Timestamp("2024-06-01")
july = pd.Timestamp("2024-07-01")
aug  = pd.Timestamp("2024-08-01")

june_active = set(panel[(panel["year_month"]==june) & (panel["active"]==1)]["customer_id"])
july_active = set(panel[(panel["year_month"]==july) & (panel["active"]==1)]["customer_id"])
aug_active  = set(panel[(panel["year_month"]==aug)  & (panel["active"]==1)]["customer_id"])

dropped_july = june_active - july_active
recovered_aug = dropped_july & aug_active
true_non_recovered = dropped_july - aug_active

print("Dropped June→July:", len(dropped_july))
print("Recovered in Aug:", len(recovered_aug))
print("True non-recovered:", len(true_non_recovered))


Dropped June→July: 33
Recovered in Aug: 23
True non-recovered: 10


In [87]:
window_start = pd.Timestamp("2024-01-01")
window_end   = pd.Timestamp("2024-06-01")

jan_jun = panel[(panel["year_month"]>=window_start) & (panel["year_month"]<=window_end)].copy()
jan_jun["inactive"] = (jan_jun["active"]==0).astype(int)

# longest inactive streak helper
def longest_inactive_streak(x):
    # x is 0/1 inactive flag in time order
    best = cur = 0
    for v in x:
        if v==1:
            cur += 1
            best = max(best, cur)
        else:
            cur = 0
    return best

drop_behavior = (
    jan_jun[jan_jun["customer_id"].isin(dropped_july)]
    .groupby("customer_id")
    .agg(
        active_months_jan_jun=("active","sum"),
        inactive_months_jan_jun=("inactive","sum"),
        longest_inactive_streak=("inactive", longest_inactive_streak),
        baseline_avg_jan_may=("net_sales", lambda s: s[jan_jun.loc[s.index,"year_month"] < june].mean()),
        avg_discount_jan_jun=("avg_discount","mean"),
        avg_category_jan_jun=("category_count","mean"),
        loyalty=("loyalty","max"),
        digital=("digital","max"),
        sales_agent_id=("sales_agent_id", "first"),
    )
    .reset_index()
)

# add June revenue and spike ratio (June / Jan–May avg)
june_rev = (
    panel[(panel["year_month"]==june) & (panel["customer_id"].isin(dropped_july))]
    .groupby("customer_id")["net_sales"].sum().rename("june_revenue").reset_index()
)

drop_behavior = drop_behavior.merge(june_rev, on="customer_id", how="left")
drop_behavior["spike_ratio"] = np.where(
    drop_behavior["baseline_avg_jan_may"]>0,
    drop_behavior["june_revenue"]/drop_behavior["baseline_avg_jan_may"],
    np.nan
)

# cohort flags
drop_behavior["dropped_july"] = 1
drop_behavior["recovered_aug"] = drop_behavior["customer_id"].isin(recovered_aug).astype(int)
drop_behavior["true_non_recovered"] = drop_behavior["customer_id"].isin(true_non_recovered).astype(int)

drop_behavior.sort_values(["true_non_recovered","june_revenue"], ascending=[False,False]).head(15)

,customer_id,active_months_jan_jun,inactive_months_jan_jun,longest_inactive_streak,baseline_avg_jan_may,avg_discount_jan_jun,avg_category_jan_jun,loyalty,digital,sales_agent_id,june_revenue,spike_ratio,dropped_july,recovered_aug,true_non_recovered
10,C00080,6,0,0,757.718,0.190772,2.000000,True,True,A015,1976.74,2.608807,1,0,1
15,C00106,6,0,0,1233.448,0.168736,2.166667,True,True,A011,1539.70,1.248289,1,0,1
12,C00086,4,2,2,811.864,0.138344,1.166667,False,False,A017,1254.21,1.544852,1,0,1
32,C00217,4,2,2,368.802,0.122544,1.333333,True,False,A015,1212.22,3.286913,1,0,1
24,C00157,4,2,1,426.542,0.130037,1.666667,False,False,A009,1106.23,2.593484,1,0,1
16,C00108,3,3,2,449.910,0.097067,1.000000,0,0,A017,665.69,1.479607,1,0,1
13,C00091,6,0,0,867.204,0.203070,1.500000,False,True,A018,537.46,0.619762,1,0,1
9,C00073,5,1,1,485.698,0.173152,1.666667,False,True,A005,496.98,1.023228,1,0,1
23,C00155,4,2,1,355.626,0.134546,1.333333,0,True,A015,272.82,0.767154,1,0,1
26,C00167,5,1,1,462.366,0.169906,1.166667,False,True,A015,129.48,0.280038,1,0,1


In [74]:
def cohort_stats(df, label):
    r = df["spike_ratio"].dropna()
    print("\n==", label, "==")
    print("n:", df["customer_id"].nunique())
    print("median spike:", np.median(r))
    print("p75 spike:", np.quantile(r, 0.75))
    print("% > 2x:", np.mean(r>2))
    print("avg active months Jan–Jun:", df["active_months_jan_jun"].mean())
    print("avg longest inactive streak:", df["longest_inactive_streak"].mean())

cohort_stats(drop_behavior, "Dropped cohort (33)")
cohort_stats(drop_behavior[drop_behavior["true_non_recovered"]==1], "True non-recovered (10)")



== Dropped cohort (33) ==
n: 33
median spike: 1.7658903333486458
p75 spike: 2.9652007210924407
% > 2x: 0.45454545454545453
avg active months Jan–Jun: 4.787878787878788
avg longest inactive streak: 1.7575757575757576

== True non-recovered (10) ==
n: 10
median spike: 1.3639481903755921
p75 spike: 2.331326344605744
% > 2x: 0.3
avg active months Jan–Jun: 4.7
avg longest inactive streak: 2.0


In [78]:
aug_active_customers = set(panel[(panel["year_month"]==aug) & (panel["active"]==1)]["customer_id"])

jan = pd.Timestamp("2024-01-01")
last6 = panel[(panel["year_month"]>=jan) & (panel["year_month"]<=aug)].copy()
last6["inactive"] = (last6["active"]==0).astype(int)

# last2 (Jul–Aug) to detect recent wobble
last2 = panel[(panel["year_month"]>=july) & (panel["year_month"]<=aug)].copy()

aug_risk = (
    last6[last6["customer_id"].isin(aug_active_customers)]
    .groupby("customer_id")
    .agg(
        active_months_last6=("active","sum"),
        inactive_months_last6=("inactive","sum"),
        longest_inactive_streak_last6=("inactive", longest_inactive_streak),
        baseline_avg_last5=("net_sales", lambda s: s[last6.loc[s.index,"year_month"]<aug].mean()),  # Mar–Jul avg
        avg_discount_last6=("avg_discount","mean"),
        avg_category_last6=("category_count","mean"),
        loyalty=("loyalty","max"),
        digital=("digital","max"),
        sales_agent_id=("sales_agent_id","first"),
    )
    .reset_index()
)

# add July revenue and Aug revenue to see rebound vs wobble
july_rev = panel[(panel["year_month"]==july) & (panel["customer_id"].isin(aug_active_customers))]\
    .groupby("customer_id")["net_sales"].sum().rename("july_revenue").reset_index()
aug_rev = panel[(panel["year_month"]==aug) & (panel["customer_id"].isin(aug_active_customers))]\
    .groupby("customer_id")["net_sales"].sum().rename("aug_revenue").reset_index()

aug_risk = aug_risk.merge(july_rev, on="customer_id", how="left").merge(aug_rev, on="customer_id", how="left")

# “spike” style feature: Aug revenue vs Mar–Jul average
aug_risk["aug_vs_baseline"] = np.where(
    aug_risk["baseline_avg_last5"]>0,
    aug_risk["aug_revenue"]/aug_risk["baseline_avg_last5"],
    np.nan
)

# define a simple “risk edge” score (you can tweak weights)
aug_risk["risk_edge_score"] = (
    0.6*aug_risk["inactive_months_last6"] +
    0.4*aug_risk["longest_inactive_streak_last6"]
)

# optional urgency bump if they had a big Aug spike (pull-forward risk next month)
aug_risk["spike_flag_aug"] = (aug_risk["aug_vs_baseline"] > 2).astype(int)

# prioritize: high value + high edge
aug_risk["priority_score"] = (
    0.7*aug_risk["baseline_avg_last5"].fillna(0) +
    0.3*aug_risk["aug_revenue"].fillna(0)
) * (1 + 0.1*aug_risk["risk_edge_score"])  # small edge boost

aug_risk.sort_values("priority_score", ascending=False).head(15)


,customer_id,active_months_last6,inactive_months_last6,longest_inactive_streak_last6,baseline_avg_last5,avg_discount_last6,avg_category_last6,loyalty,digital,sales_agent_id,july_revenue,aug_revenue,aug_vs_baseline,risk_edge_score,spike_flag_aug,priority_score
24,C00089,7,1,1,866.698571,0.154908,1.500,True,False,A006,1383.58,3219.03,3.714129,1.0,1,1729.63780
0,C00002,7,1,1,838.005714,0.157714,2.000,True,False,A015,1233.21,2104.33,2.511117,1.0,1,1339.69330
36,C00131,4,4,3,192.765714,0.092272,0.750,True,True,A016,0.00,2650.86,13.751719,3.6,1,1265.06384
48,C00173,6,2,1,803.832857,0.163592,1.625,0,True,A006,1300.33,1315.60,1.636659,1.6,0,1110.54108
15,C00060,5,3,2,446.915714,0.132061,1.250,0,0,A011,0.00,1845.10,4.128519,2.6,1,1091.62746
61,C00219,8,0,0,898.905714,0.177606,2.000,True,True,A001,83.92,1426.49,1.586918,0.0,0,1057.18100
34,C00126,6,2,1,836.978571,0.134776,1.750,True,False,A014,1000.14,1039.84,1.242374,1.6,0,1041.49092
43,C00154,5,3,2,422.775714,0.116869,1.000,False,False,A004,1476.98,1736.85,4.108207,2.6,1,1029.41748
62,C00220,5,3,3,699.862857,0.134043,1.375,True,False,A009,526.60,792.31,1.132093,3.0,0,945.87610
53,C00193,6,2,1,502.317143,0.145469,1.500,True,True,A017,671.17,1533.64,3.053131,1.6,1,941.58824


In [79]:
import pandas as pd
import numpy as np

panel = panel.copy()
panel["year_month"] = pd.to_datetime(panel["year_month"])
panel = panel.sort_values(["customer_id","year_month"])

aug = pd.Timestamp("2024-08-01")

aug_active_customers = set(
    panel[(panel["year_month"]==aug) & (panel["active"]==1)]["customer_id"]
)

print("Aug-active customers:", len(aug_active_customers))


Aug-active customers: 63


In [81]:
start = pd.Timestamp("2024-01-01")
end   = pd.Timestamp("2024-08-01")

jan_aug = panel[(panel["year_month"]>=start) & (panel["year_month"]<=end)].copy()
jan_aug["inactive"] = (jan_aug["active"]==0).astype(int)

def longest_inactive_streak(x):
    best = cur = 0
    for v in x:
        if v==1:
            cur += 1
            best = max(best, cur)
        else:
            cur = 0
    return best

# customer-month net sales time series helper for volatility
def cv_nonzero(s):
    s = s[s>0]
    if len(s) <= 1:
        return np.nan
    mu = s.mean()
    return np.nan if mu==0 else s.std()/mu

health = (
    jan_aug[jan_aug["customer_id"].isin(aug_active_customers)]
    .groupby("customer_id")
    .agg(
        active_months_jan_aug=("active","sum"),
        inactive_months_jan_aug=("inactive","sum"),
        longest_inactive_streak=("inactive", longest_inactive_streak),

        total_sales_jan_aug=("net_sales","sum"),
        avg_sales_active_month=("net_sales", lambda s: s[s>0].mean()),
        median_sales_active_month=("net_sales", lambda s: s[s>0].median()),
        sales_cv=("net_sales", cv_nonzero),

        avg_discount=("avg_discount","mean"),
        avg_category=("category_count","mean"),

        loyalty=("loyalty","max"),
        digital=("digital","max"),
        sales_agent_id=("sales_agent_id","first"),
    )
    .reset_index()
)

# add last purchase month (recency within Jan–Aug window)
last_purchase = (
    jan_aug[(jan_aug["customer_id"].isin(aug_active_customers)) & (jan_aug["active"]==1)]
    .groupby("customer_id")["year_month"].max()
    .rename("last_active_month")
    .reset_index()
)

health = health.merge(last_purchase, on="customer_id", how="left")
health.head()


,customer_id,active_months_jan_aug,inactive_months_jan_aug,longest_inactive_streak,total_sales_jan_aug,avg_sales_active_month,median_sales_active_month,sales_cv,avg_discount,avg_category,loyalty,digital,sales_agent_id,last_active_month
0,C00002,7,1,1,7970.37,1138.624286,1092.22,0.475170,0.157714,2.000,True,False,A015,2024-08-01
1,C00006,5,3,2,2599.54,519.908000,533.18,0.247042,0.134417,1.125,False,False,A011,2024-08-01
2,C00007,3,5,4,3598.53,1199.510000,1312.81,0.210652,0.084570,0.750,0,0,A006,2024-08-01
3,C00013,5,3,2,5425.10,1085.020000,1022.65,0.333105,0.123258,1.500,True,True,A006,2024-08-01
4,C00014,4,4,1,2067.80,516.950000,490.18,0.298011,0.117647,0.875,0,0,A005,2024-08-01


In [82]:
july = pd.Timestamp("2024-07-01")

july_active = set(panel[(panel["year_month"]==july) & (panel["active"]==1)]["customer_id"])

health["inactive_in_july"] = (~health["customer_id"].isin(july_active)).astype(int)
health["recent_wobble_flag"] = health["inactive_in_july"]  # Jul=0, Aug=1 pattern

health["multi_gap_flag"] = (health["inactive_months_jan_aug"] >= 2).astype(int)
health["streak_flag"] = (health["longest_inactive_streak"] >= 2).astype(int)
health["high_discount_flag"] = (health["avg_discount"] >= 0.20).astype(int)


In [83]:
june = pd.Timestamp("2024-06-01")

baseline_jan_jun = (
    panel[(panel["year_month"]>=start) & (panel["year_month"]<=june) &
          (panel["customer_id"].isin(aug_active_customers))]
    .groupby("customer_id")["net_sales"].mean()
    .rename("baseline_avg_jan_jun")
    .reset_index()
)

aug_rev = (
    panel[(panel["year_month"]==aug) & (panel["customer_id"].isin(aug_active_customers))]
    .groupby("customer_id")["net_sales"].sum()
    .rename("aug_revenue")
    .reset_index()
)

health = health.merge(baseline_jan_jun, on="customer_id", how="left").merge(aug_rev, on="customer_id", how="left")

health["aug_spike_ratio"] = np.where(
    health["baseline_avg_jan_jun"]>0,
    health["aug_revenue"]/health["baseline_avg_jan_jun"],
    np.nan
)

health["aug_spike_flag_2x"] = (health["aug_spike_ratio"] > 2).astype(int)


In [84]:
health["risk_edge_score"] = (
    1.0*health["recent_wobble_flag"] +
    0.6*health["multi_gap_flag"] +
    0.6*health["streak_flag"] +
    0.3*health["high_discount_flag"]
)

# prioritize by value * (1 + edge)
health["priority_score"] = health["avg_sales_active_month"].fillna(0) * (1 + 0.4*health["risk_edge_score"])

watchlist = health.sort_values("priority_score", ascending=False).copy()
watchlist["risk_rank"] = range(1, len(watchlist)+1)


In [85]:
health.to_csv("CustomerHealth_AugActive.csv", index=False)
watchlist.head(30).to_csv("AugActive_RiskList_Top30.csv", index=False)

print("Exported: CustomerHealth_AugActive.csv")
print("Exported: AugActive_RiskList_Top30.csv")


Exported: CustomerHealth_AugActive.csv
Exported: AugActive_RiskList_Top30.csv


In [88]:
import pandas as pd
import numpy as np

panel = panel.copy()
panel["year_month"] = pd.to_datetime(panel["year_month"])
panel = panel.sort_values(["customer_id","year_month"])

jan = pd.Timestamp("2024-01-01")
jun = pd.Timestamp("2024-06-01")
jul = pd.Timestamp("2024-07-01")
aug = pd.Timestamp("2024-08-01")

aug_active = set(panel[(panel["year_month"]==aug) & (panel["active"]==1)]["customer_id"])
print("Aug-active:", len(aug_active))


Aug-active: 63


In [90]:
jan_aug = panel[(panel["year_month"]>=jan) & (panel["year_month"]<=aug)].copy()
jan_aug["inactive"] = (jan_aug["active"]==0).astype(int)

def longest_inactive_streak(x):
    best = cur = 0
    for v in x:
        if v==1:
            cur += 1
            best = max(best, cur)
        else:
            cur = 0
    return best

cadence = (
    jan_aug[jan_aug["customer_id"].isin(aug_active)]
    .groupby("customer_id")
    .agg(
        active_months_jan_aug=("active","sum"),
        inactive_months_jan_aug=("inactive","sum"),
        longest_inactive_streak=("inactive", longest_inactive_streak),
        avg_discount_jan_aug=("avg_discount","mean"),
        avg_category_jan_aug=("category_count","mean"),
        loyalty=("loyalty","max"),
        digital=("digital","max"),
        sales_agent_id=("sales_agent_id","first"),
    )
    .reset_index()
)

# sanity check: must equal 8 months
assert ((cadence["active_months_jan_aug"] + cadence["inactive_months_jan_aug"]) == 8).all()


In [91]:
# baseline typical month before Aug
baseline = (
    panel[(panel["year_month"]>=jan) & (panel["year_month"]<=jul) &
          (panel["customer_id"].isin(aug_active))]
    .groupby("customer_id")["net_sales"]
    .mean()
    .rename("baseline_avg_jan_jul")
    .reset_index()
)

aug_rev = (
    panel[(panel["year_month"]==aug) & (panel["customer_id"].isin(aug_active))]
    .groupby("customer_id")["net_sales"]
    .sum()
    .rename("aug_revenue")
    .reset_index()
)

july_rev = (
    panel[(panel["year_month"]==jul) & (panel["customer_id"].isin(aug_active))]
    .groupby("customer_id")["net_sales"]
    .sum()
    .rename("july_revenue")
    .reset_index()
)

aug_tbl = cadence.merge(baseline, on="customer_id", how="left") \
                 .merge(aug_rev, on="customer_id", how="left") \
                 .merge(july_rev, on="customer_id", how="left")

aug_tbl["aug_spike_ratio"] = np.where(
    aug_tbl["baseline_avg_jan_jul"]>0,
    aug_tbl["aug_revenue"]/aug_tbl["baseline_avg_jan_jul"],
    np.nan
)


In [92]:
# 1) cadence weakness (same logic you wrote)
aug_tbl["cadence_weak_flag"] = (
    (aug_tbl["active_months_jan_aug"] <= 5) |  # <=5 out of 8 months active (adjustable)
    (aug_tbl["longest_inactive_streak"] >= 2)
).astype(int)

# 2) spike flags (same logic)
aug_tbl["spike_ge_2_flag"] = (aug_tbl["aug_spike_ratio"] >= 2).astype(int)
aug_tbl["spike_lt_1_flag"] = (aug_tbl["aug_spike_ratio"] < 1).astype(int)

# 3) recent wobble: July inactive but Aug active
# If july_revenue is NaN or 0, treat as inactive in July
aug_tbl["july_inactive_flag"] = (aug_tbl["july_revenue"].fillna(0) <= 0).astype(int)

# 2x2 segments (very good for presentation)
def segment(row):
    if row["cadence_weak_flag"]==1 and row["spike_ge_2_flag"]==0:
        return "Unstable + Normal (true-loss risk)"
    if row["cadence_weak_flag"]==1 and row["spike_ge_2_flag"]==1:
        return "Unstable + Spike (discount/bulk dependent)"
    if row["cadence_weak_flag"]==0 and row["spike_ge_2_flag"]==1:
        return "Stable + Spike (pull-forward risk)"
    return "Stable + Normal (healthy)"

aug_tbl["risk_segment"] = aug_tbl.apply(segment, axis=1)


In [93]:
# priority score: value * (risk adjustment)
aug_tbl["risk_multiplier"] = (
    1
    + 0.6*aug_tbl["cadence_weak_flag"]
    + 0.3*aug_tbl["july_inactive_flag"]
    + 0.2*aug_tbl["spike_lt_1_flag"]     # cooling demand
    - 0.1*(aug_tbl["aug_spike_ratio"]>=3)  # extreme spike may be bulk; reduce panic
)

aug_tbl["priority_score"] = aug_tbl["baseline_avg_jan_jul"].fillna(0) * aug_tbl["risk_multiplier"]
aug_tbl = aug_tbl.sort_values("priority_score", ascending=False)
aug_tbl["risk_rank"] = range(1, len(aug_tbl)+1)


In [94]:
aug_tbl.to_csv("AugustActive_RiskFingerprint_JanAug.csv", index=False)

# A clean Top 30 watchlist
aug_tbl.head(30).to_csv("AugustActive_Watchlist_Top30.csv", index=False)

print("Exported AugustActive_RiskFingerprint_JanAug.csv")
print("Exported AugustActive_Watchlist_Top30.csv")


Exported AugustActive_RiskFingerprint_JanAug.csv
Exported AugustActive_Watchlist_Top30.csv
